# Import the package

In [14]:
import math

import numpy as np
from tqdm.notebook import tqdm

# Problem 1

In [12]:
def get_perimeter(x1, x2, x3, y1, y2, y3):
    return ((x1-x2)**2 + (y1 - y2)**2)**0.5 + ((x1-x3)**2 + (y1-y3)**2)**0.5 + ((x2-x3)**2 + (y2-y3)**2)**0.5

def get_area(x1, x2, x3, y1, y2, y3):
    return 0.5*abs((x1*y2-x2*y1)+(x2*y3-x3*y2)+(x3*y1-x1*y3))

def is_obtuse(x1, x2, x3, y1, y2, y3):
    vector1 = (x2-x1)*(x3-x1) + (y2-y1)*(y3-y1)
    vector2 = (x3-x2)*(x1-x2) + (y3-y2)*(y1-y2)
    vector3 = (x1-x3)*(x2-x3) + (y1-y3)*(y2-y3)
    if vector1 > 0 and vector2 > 0 and vector3 > 0:
        return 0
    else:
        return 1

In [13]:
perimeters = []
areas = []
obtuses = []
for i in tqdm(range(0,5000)):
    x1, y1, x2, y2, x3, y3 = tuple(np.random.rand(6))
    perimeters.append(get_perimeter(x1, x2, x3, y1, y2, y3))
    areas.append(get_area(x1, x2, x3, y1, y2, y3))
    obtuses.append(is_obtuse(x1, x2, x3, y1, y2, y3))

average_perimeter = np.mean(perimeters)
average_area = np.mean(areas)
probability_obtuse = np.mean(obtuses)

print("The average perimeter of 5000 Monte-Carlo Simulations is {}".format(average_perimeter))
print("The average area of 5000 Monte-Carlo Simulations is {}".format(average_area))
print("The probability that this triangle is obtuse of 5000 Monte-Carlo Simulations is {}".format(probability_obtuse))

  0%|          | 0/5000 [00:00<?, ?it/s]

The average perimeter of 5000 Monte-Carlo Simulations is 1.5659643689857432
The average area of 5000 Monte-Carlo Simulations is 0.0759041132541655
The probability that this triangle is obtuse of 5000 Monte-Carlo Simulations is 0.7306


**Result**:
The average perimeter of 5000 Monte-Carlo Simulations is 1.5659643689857432
The average area of 5000 Monte-Carlo Simulations is 0.0759041132541655
The probability that this triangle is obtuse of 5000 Monte-Carlo Simulations is 0.7306

# Problem 2

In [15]:
def discriminant(a, b, c):
    return b*b - 4*a*c

def have_real_roots(a, b, c):
    if discriminant(a, b, c) >= 0:
        return 1
    else:
        return 0
    
def solve_equation(a, b, c):
    D = discriminant(a, b, c)
    root1 = (-b + math.sqrt(D)) / (2*a)
    root2 = (-b - math.sqrt(D)) / (2*a)
    return root1, root2

In [25]:
have_real_roots_list = []
for i in tqdm(range(5000)):
    a, b, c = tuple(np.random.rand(3))
    have_real_roots_list.append(have_real_roots(a, b, c))
    
roots1 = []
roots2 = []
while True:
    a, b, c = tuple(np.random.rand(3))
    if discriminant(a, b, c) >= 0:
        root1, root2 = solve_equation(a, b, c)
        roots1.append(root1)
        roots2.append(root2)
        if len(roots1) > 5000:
            break

print("The probability that both roots of this equation are real of 5000 Monte-Carlo Simulations is {}".format(np.mean(have_real_roots_list)))
print("The root1 of this equation of 5000 Monte-Carlo Simulations is {}".format(np.mean(roots1)))
print("The root2 of this equation of 5000 Monte-Carlo Simulations is {}".format(np.mean(roots2)))

  0%|          | 0/5000 [00:00<?, ?it/s]

The probability that both roots of this equation are real of 5000 Monte-Carlo Simulations is 0.2546
The root1 of this equation of 5000 Monte-Carlo Simulations is -0.5767142878400107
The root2 of this equation of 5000 Monte-Carlo Simulations is -14.32722766903135


**Result:**
The probability that both roots of this equation are real of 5000 Monte-Carlo Simulations is 0.2546
The root1 of this equation of 5000 Monte-Carlo Simulations is -0.5767142878400107
The root2 of this equation of 5000 Monte-Carlo Simulations is -14.32722766903135

# Problem 3

In [26]:
def distance_point_to_line_segment(A, B, C):
    AB = B - A
    AC = C - A
    t = np.dot(AC, AB) / np.dot(AB, AB)

    if t < 0.0:
        nearest_point = A
    elif t > 1.0:
        nearest_point = B
    else:
        nearest_point = A + t * AB

    return np.linalg.norm(C - nearest_point)

def intersect(A, B, C, R):
    if distance_point_to_line_segment(A, B, C) <= R:
        return 1
    else:
        return 0

In [30]:
is_intersect = []
for i in tqdm(range(5000)):
    end1 = np.random.rand(2)
    end2 = np.random.rand(2)
    center = np.random.rand(2)
    radius = np.random.rand()
    is_intersect.append(intersect(end1, end2, center, radius))
    
print('The probability that the line intersects the circle of 5000 Monte-Carlo Simulations is {}'.format(np.mean(is_intersect)))

  0%|          | 0/5000 [00:00<?, ?it/s]

The probability that the line intersects the circle of 5000 Monte-Carlo Simulations is 0.6378


**Result:**
The probability that the line intersects the circle of 5000 Monte-Carlo Simulations is 0.6378